source/retryable-writes/retryable-writes.rst

```diff
@@ -433,66 +435,76 @@ Consider the following pseudo-code:
 
     /* If the server does not support retryable writes, execute the write as if
      * retryable writes are not enabled. */
-    if ( ! isRetryableWritesSupported()) {
+    if ( ! isRetryableWritesSupported(server)) {
       return executeCommand(server, command);
     }
 
     /* Incorporate lsid and txnNumber fields into the command document. These
      * values will be derived from the implicit or explicit session object. */
     retryableCommand = addTransactionIdToCommand(command, session);
 
-    /* NetworkException and NotMasterException are both retryable errors. If
-     * caught, remember the exception, update SDAM accordingly, and proceed with
-     * retrying the operation.
-     *
-     * IllegalOperation (code 20) with errmsg starting with "Transaction
-     * numbers" MUST be re-raised with an actionable error message. */
-    try {
-      return executeCommand(server, retryableCommand);
-    } catch (NetworkException originalError) {
-      updateTopologyDescriptionForNetworkError(server, originalError);
-    } catch (NotMasterException originalError) {
-      updateTopologyDescriptionForNotMasterError(server, originalError);
-    } catch (IllegalOperation originalError) {
-      if ( originalError.code == 20 && originalError.errmsg.startsWith("Transaction numbers") ) {
-        originalError.errmsg = "This MongoDB deployment does not support retryable...";
-        throw originalError;
+    Exception previousError = null;
+    int numSocketTimeouts = 0;
+    while true {
+      /* NetworkException and NotMasterException are both retryable errors. If
+       * caught, remember the exception, update SDAM accordingly, and proceed with
+       * retrying the operation.
+       *
+       * IllegalOperation (code 20) with errmsg starting with "Transaction
+       * numbers" MUST be re-raised with an actionable error message.
+       *
+       * Exceptions that originate from the driver (e.g. no socket available
+       * from the connection pool) are treated as fatal. Any such exception
+       * that occurs on the original attempt is propagated as-is. On retries,
+       * the error from the previous attempt is raised as it will be more
+       * relevant for the user. */
+      try {
+        return executeCommand(server, retryableCommand);
+      } catch (NetworkException networkError) {
+        updateTopologyDescriptionForNetworkError(server, networkError);
+
+        // Check if we've reached the socket timeout limit and terminate the
+        // loop.
+        if (networkError is NetworkTimeoutException) {
+          numSocketTimeouts++;
+          if (numSocketTimeouts == 2) {
+            throw networkError;
+          }
+        }
+
+        previousError = networkError;
+
+      } catch (NotMasterException notMasterError) {
```

Comment:
This will be a merge conflict with https://github.com/mongodb/specifications/pull/974. I suppose you'll rebase locally before merging to catch this, though.